### PR TITLE
Move LMDBStore to shared subproject and reuse it in block-storage

### PR DESCRIPTION
**Pull request by itegulov**
## Overview
Title says it all

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3251

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: rspace/src/main/scala/coop/rchain/rspace/history/Store.scala [hunk 1]
```diff
@@ -38,91 +37,51 @@ object StoreInstances {
...
-    withReadTxnF { txn =>
-      dbi.get(txn, key)
-    }.map(v => Option(v))
-
-  def put(key: Blake2b256Hash, value: BitVector): F[Unit] = {
-    val directKey   = key.bytes.toDirectByteBuffer
-    val directValue = value.toByteVector.toDirectByteBuffer
-    put(directKey, directValue)
-  }
+    val dbi   = env.openDbi("db", MDB_CREATE)
```
Comment (sebtomba) [inline rspace/src/main/scala/coop/rchain/rspace/history/Store.scala]: @itegulov Can you refactor
```scala
def lmdbStore[F[_]: Sync](config: StoreConfig): Store[F] =
```
to
```scala
def lmdbStore[F[_]: Sync](config: StoreConfig): F[Store[F]]
```
?

**Review by sebtomba — APPROVED** [review]
LGTM. But please consider my request.

File: shared/src/main/scala/coop/rchain/lmdb/LMDBStore.scala [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  def get(key: ByteBuffer): F[Option[ByteBuffer]] =
+    withReadTxnF { txn =>
+      dbi.get(txn, key)
+    }.map(v => Option(v))
+
+  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
+  def put(key: ByteBuffer, value: ByteBuffer): F[Unit] =
+    withWriteTxnF { txn =>
+      if (dbi.put(txn, key, value)) {
```
Comment (lukasz-golebiewski) [inline shared/src/main/scala/coop/rchain/lmdb/LMDBStore.scala]: I know this code was merely moved but wouldn't :
```
 if (!dbi.put(txn, key, value)) {
        throw new RuntimeException("was not able to put data")
      }
```
be more readable?

File: rspace/src/test/scala/coop/rchain/rspace/history/HistoryRepositoryGenerativeSpec.scala [hunk 3]
```diff
@@ -46,13 +46,19 @@ class LMDBHistoryRepositoryGenerativeSpec
     )
 
   def lmdbHistoryStore =
-    HistoryStoreInstances.historyStore(StoreInstances.lmdbStore[Task](lmdbConfig))
+    HistoryStoreInstances.historyStore(
+      StoreInstances.lmdbStore[Task](lmdbConfig).runSyncUnsafe(Duration.Inf)
```
Comment (lukasz-golebiewski) [inline rspace/src/test/scala/coop/rchain/rspace/history/HistoryRepositoryGenerativeSpec.scala]: Is it a lot of effort to get rid of the `runSyncUsafe`s?

File: rspace/src/main/scala/coop/rchain/rspace/history/Store.scala [hunk 4]
```diff
@@ -31,98 +30,60 @@ final case class StoreConfig(
...
+
+      @SuppressWarnings(Array("org.wartremover.warts.Throw"))
+      private[this] def putIfAbsent(
+          txn: Txn[ByteBuffer],
+          key: ByteBuffer,
+          value: ByteBuffer
+      ): Unit =
+        Option(dbi.get(txn, key)) match {
+          case None =>
+            if (dbi.put(txn, key, value)) {
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/history/Store.scala]: This code looks similar to what's already in `LMDBStore`. Why not re-use?

File: rspace/src/test/scala/coop/rchain/rspace/history/HistoryRepositoryGenerativeSpec.scala (hunk 3 above)
Comment (itegulov) [reply rspace/src/test/scala/coop/rchain/rspace/history/HistoryRepositoryGenerativeSpec.scala]: Done

File: shared/src/main/scala/coop/rchain/lmdb/LMDBStore.scala (hunk 2 above)
Comment (itegulov) [reply shared/src/main/scala/coop/rchain/lmdb/LMDBStore.scala]: Good point

File: rspace/src/main/scala/coop/rchain/rspace/history/Store.scala (hunk 4 above)
Comment (itegulov) [reply rspace/src/main/scala/coop/rchain/rspace/history/Store.scala]: The code in `LMDBStore` encapsulates transactions inside of its methods. This method, on the other hand, is being used in rspace-specific `put` which runs multiple puts during a single transaction.

**Review by lukasz-golebiewski — APPROVED** [review]

**Pull request merged**